Refactor to generalize services

darinkrauss opened the pull request:
- Add Service and ServiceUI protocols
- Implement common ServiceTableViewController and ServiceViewController for simple services
- Add Analytics, Logging, and RemoteData protocols
- Add DiagnosticLog
- Move GlucoseThreshold and TempBasalRecommendation from Loop
- Implement MockService to record events from Analytics, Logging, and RemoteData protocols

https://tidepool.atlassian.net/browse/LOOP-2
https://tidepool.atlassian.net/browse/LOOP-11
https://tidepool.atlassian.net/browse/LOOP-13
https://tidepool.atlassian.net/browse/LOOP-12
https://tidepool.atlassian.net/browse/LOOP-10

darinkrauss committed:
Refactor to generalize services

- Add Service and ServiceUI protocols
- Implement common ServiceTableViewController and ServiceViewController for simple services
- Add Analytics, Logging, and RemoteData protocols
- Add DiagnosticLog
- Move GlucoseThreshold and TempBasalRecommendation from Loop
- Implement MockService to record events from Analytics, Logging, and RemoteData procotols

nhamming reviewed (approved):
LGTM

ps2 commented on LoopKit/RemoteData.swift:
```diff
@@ -0,0 +1,37 @@
+import Foundation
+
+
+public protocol RemoteData {
```
`RemoteDataService` would help indicate that this is a protocol specifying the interface for a service, rather than the definition of a data object, or struct like interface, as `RemoteData` suggests.

darinkrauss replied:
Okay, will do.

darinkrauss replied:
Done.

ps2 commented on LoopKit/Service.swift:
```diff
@@ -0,0 +1,107 @@
+    var serviceDelegate: ServiceDelegate? { get set }
+
+    /// Does the service have a valid configuration?
+    var hasValidConfiguration: Bool { get }
```
Configuration should generally be enforced during setup.  If the user starts to add a service, and doesn't complete entering username/password/etc, then we don't add the service. This kind of checking shouldn't be Loop's responsibility.

ps2 replied:
I think this will go away when the comments about ServiceTableViewController.swift are addressed, so you can mostly ignore the above thinking.

darinkrauss replied:
I'll rename `hasValidConfiguration` to just `hasConfiguration` to make it distinct from `verifyConfiguration`. That is, the first step is to check whether the service has a configuration and then allow the service to verify the configuration (eg. Nightscout checks to see if the URL and secret work) when the user taps Done.

darinkrauss replied:
Done.

ps2 commented on LoopKit/Service.swift:
```diff
@@ -0,0 +1,107 @@
+    ///
+    /// - Parameters:
+    ///     - completion: A closure called once upon completion with any error.
+    func verifyConfiguration(completion: @escaping (Error?) -> Void)
```
This seems to provide the same information as above?

darinkrauss replied:
See above comment.

darinkrauss replied:
Done.

ps2 commented on LoopKit/Service.swift:
```diff
@@ -0,0 +1,107 @@
+    ///
+    /// - Parameters:
+    ///     - completion: A closure called once upon completion.
+    func notifyCreated(completion: @escaping () -> Void)
```
Why would a service need to be notified about its own creation?

ps2 replied:
Again, I think this will go away when the comments about ServiceTableViewController.swift are addressed, so you can mostly ignore this question.

ps2 commented on LoopKit/Service.swift:
```diff
@@ -0,0 +1,107 @@
+    ///
+    /// - Parameters:
+    ///     - completion: A closure called once upon completion.
+    func notifyDeleted(completion: @escaping () -> Void)
```
The ServiceDelegate has a serviceDeleted method as well. Who has the responsibility for deleting the service? Typically in DeviceManagers we let the service itself trigger deletion and then notify the delegate.  I could see this going the other way, but it should be one or the other, not both.

ps2 replied:
Same: When comments about ServiceTableViewController.swift are addressed, this will probably be gone, so no need to answer these questions.

ps2 commented on LoopKitUI/View Controllers/ServiceTableViewController.swift:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    private func notifyCreated() {
+        service.notifyCreated {
```
I now see this is probably the reason you've added these methods to the Service protocol.  They make more sense as part of a ServiceTableViewControllerDelegate protocol.  Not all services will use this tableview, I imagine.  It's fine to have this VC here as a convenience, but it should have its own delegate, and the methods for it not mixed in with the Service protocol.  That allows it to be used independently.

darinkrauss replied:
I believe I have straightened out the notification and delegate issues now. Take a look and let me know what you think.

ps2 commented on LoopKit/Analytics.swift:
```diff
@@ -0,0 +1,16 @@
+import Foundation
+
+
+public protocol Analytics {
```
AnalyticsService?

darinkrauss replied:
Will do.

darinkrauss replied:
Done.

ps2 commented on LoopKit/Logging.swift:
```diff
@@ -0,0 +1,37 @@
+import os.log
+
+
+public protocol Logging {
```
LoggingService?

darinkrauss replied:
Will do.

darinkrauss replied:
Done.

ps2 commented on LoopKit/RemoteData.swift:
```diff
@@ -0,0 +1,37 @@
+
+    func upload(glucoseValues values: [GlucoseValue], sensorState: SensorDisplayable?)
+
+    func upload(pumpEvents events: [PersistedPumpEvent], fromSource source: String, completion: @escaping (Result<[URL], Error>) -> Void)
```
At some point I would imagine that this (and the other datatypes) would change to be query initiated by the service to ask the delegate for events, passing a query anchor that this service manages?  What is in this PR is fine as it matches the current paradigm well. Just checking that we're on the same page for eventual implementation.

darinkrauss replied:
Yes, the next chunk of work is to allow the services to "pull" the events from the delegate using anchors and such. This is really just an intermediate (but fully functional) step on the way to the final solution.

ps2 reviewed (changes requested):
ServiceTableViewController needs its own delegate protocol, and methods moved from Service protocol. A few naming suggestions, and a question.

darinkrauss committed:
Fix bug with MockService view controller

darinkrauss committed:
Rename new service based protocols

- Rename Analytics protocol to AnalyticsService
- Rename Logging protocol to LoggingService
- Rename RemoteData protocol to RemoteDataService
- Rename SharedLogging to SharedLoggingService
- Minor refactor of tests to incorporate renames

darinkrauss committed:
Rename Service hasValidConfiguration to hasConfiguration for PR review

darinkrauss committed:
Update service completion and notification delegate from PR review

darinkrauss committed:
Remove extraneous whitespace

ps2 commented on LoopKit/Service.swift:
```diff
@@ -0,0 +1,82 @@
+    ///
+    /// - Parameters:
+    ///     - service: The service created.
+    func notifyServiceCreated(_ service: Service)
```
Any non UI client dealing with a Service shouldn't need to know if the service was created (as it will have instantiated the object itself). 

The ServiceUI protocol should provide a method for getting a Setup ViewController and that view controller should have a delegate that can notify if the setup was complete (passing back the configured Service) or failed.

darinkrauss replied:
Okay, perhaps the problem here is one of naming. The purpose of this particular delegate (regardless of the name) is to tell the `ServicesManager` (the thing that keeps track of the currently active services) that a service was created/updated/delete and to persist that change. Does that clear up why this is here? Can you propose a better name that makes this more clear?

ps2 replied:
Ok, yeah, I missed that connection. I still think the existing pattern of separating the creation interactions out into the UI portion of the integration is cleaner, as it isolates those interactions to the objects that deal with that responsibility. It looks like you *are* following the pattern of having a SetupViewController. Having that vc conform to a protocol that allows for a setupDelegate will allow the completion to be handled by whatever object is dealing with setup.

darinkrauss replied:
Okay, have a look now. I *think* this is what you are looking for. If not, can provide me exactly what you are looking for?

ps2 commented on LoopKit/Service.swift:
```diff
@@ -0,0 +1,82 @@
+    ///
+    /// - Parameters:
+    ///     - service: The service updated.
+    func notifyServiceUpdated(_ service: Service)
```
The 'notify' prefix for this method and the below deletion method is unclear who is being notified. `func serviceDidUpdateState(_ service: Service)` would follow the existing convention for a device manager notifying its delegate about state updates.

ps2 commented on LoopKit/Service.swift:
```diff
@@ -0,0 +1,82 @@
+    ///
+    /// - Parameters:
+    ///     - service: The service deleted.
+    func notifyServiceDeleted(_ service: Service)
```
`func serviceWillDeactivate(_ service: Service)` (from PumpManager), or `func serviceWantsDeletion(_ service: Service)` (from CGMManager) would follow the existing conventions. I think the latter is more clear. We could update PumpManager to follow this too.

ps2 commented on LoopKit/Service.swift:
```diff
@@ -0,0 +1,82 @@
+
+}
+
+public protocol Service: DeviceManager {
```
All the methods in this block are serving a role of facilitating setup of the Service, and are only used in relationship with ServiceTableViewController. I would propose that they be moved to a ServiceTableViewControllerDelegate protocol.

ps2 commented on LoopKit/LoggingService.swift:
```diff
@@ -0,0 +1,35 @@
+    ///   - category: The category for the message. Typically the class or extension name.
+    ///   - type: The type of the message. One of OSLogType.
+    ///   - args: Optional arguments to be substituted into the string.
+    func log(_ message: StaticString, subsystem: String, category: String, type: OSLogType, _ args: [CVarArg])
```
I think we're going to want to revisit this. Maybe I'll try tackling it as part of https://tidepool.atlassian.net/browse/LOOP-354

We don't want to be sending up every log entry to the server in a separate network connection.  They need to be batched somehow. And we may need a facility for logging more complex objects, like dictionaries. `bigLog(name: .doseComputationCompleted, data: loopDecisionDataDictionary)`, or something like that.

ps2 replied:
For now; will accept since it's just refactoring existing behavior.

ps2 reviewed (changes requested):
I think the Service protocol needs to be cleaned up to remove functionality around service setup. The current block and tackle UI, though used by most (all?) of the current services, is not a coupling we should enforce as we look towards a more sophisticated service setup UI.  It's fine keeping that block and tackle UI for the existing services, but we shouldn't require future services to use it.

nhamming reviewed (approved):
+1

darinkrauss committed:
Refactor service setup delegate from PR comments

darinkrauss committed:
Fix broken tests

darinkrauss committed:
Merge branch 'dev' into services

darinkrauss committed:
Remove unnecessary elements of Service protocol

ps2 reviewed (approved):
LGTM!
